### PR TITLE
Bugfix/update trainer properties

### LINKED ISSUE #3974: [Bug]: Late update of Trainer `current_epoch` property for `LightningDataModule`
## 🐛 Bug

 Late update of Trainer `current_epoch` property for `LightningDataModule` object. 

### To Reproduce
The below code reproduces the issue:

Please check for the print logs for the `current_epoch` number in `train_dataloader`. 

#### Code sample
```
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchvision import transforms
from torchvision.datasets import MNIST
from torch.utils.data import random_split, DataLoader

import pytorch_lightning as pl


class LitModel(pl.LightningModule):

    def __init__(self, channels, width, height, num_classes, hidden_size=64, learning_rate=2e-4):

        super().__init__()

        # We take in input dimensions as parameters and use those to dynamically build model.
        self.channels = channels
        self.width = width
        self.height = height
        self.num_classes = num_classes
        self.hidden_size = hidden_size
        self.learning_rate = learning_rate

        self.model = nn.Sequential(
            nn.Flatten(),
            nn.Linear(channels * width * height, hidden_size),
            nn.ReLU(),
            nn.Dropout(0.1),
            nn.Linear(hidden_size, hidden_size),
            nn.ReLU(),
            nn.Dropout(0.1),
            nn.Linear(hidden_size, num_classes))

    def forward(self, x):
        x = self.model(x)
        return F.log_softmax(x, dim=1)

    def training_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        loss = F.nll_loss(logits, y)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
        return optimizer


class MNISTDataModule(pl.LightningDataModule):

    def __init__(self, data_dir: str = './'):
        super().__init__()
        self.data_dir = data_dir
        self.transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])

        # self.dims is returned when you call dm.size()
        # Setting default dims here because we know them.
        # Could optionally be assigned dynamically in dm.setup()
        self.dims = (1, 28, 28)
        self.num_classes = 10

    def prepare_data(self):
        # download
        MNIST(self.data_dir, train=True, download=True)
        MNIST(self.data_dir, train=False, download=True)

    def setup(self, stage=None):

        # Assign train/val datasets for use in dataloaders
        if stage == 'fit' or stage is None:
            mnist_full = MNIST(self.data_dir, train=True, transform=self.transform)
            self.mnist_train, self.mnist_val = random_split(mnist_full, [55000, 5000])

        # Assign test dataset for use in dataloader(s)
        if stage == 'test' or stage is None:
            self.mnist_test = MNIST(self.data_dir, train=False, transform=self.transform)

    def train_dataloader(self):
        print('\n----------------------------')
        print(f'Current epoch: {self.trainer.current_epoch}')
        print('----------------------------')
        if self.trainer.current_epoch > 2:
            return DataLoader(self.mnist_train, batch_size=32)
        else:
            return DataLoader(self.mnist_train, batch_size=32)


# Init DataModule
dm = MNISTDataModule()
# Init model from datamodule's attributes
model = LitModel(*dm.size(), dm.num_classes)
# Init trainer
trainer = pl.Trainer(max_epochs=5, progress_bar_refresh_rate=20, gpus=1, reload_dataloaders_every_epoch=True)
# Pass the datamodule as arg to trainer.fit to override model hooks :)
trainer.fit(model, dm)
```

#### logs

*Note*: `current_epoch` is 0 two time, which is indeed due to the late update of this property
```
----------------------------
Current epoch: 0
----------------------------
/home/nthere/2020/pytorch-lightning/pytorch_lightning/utilities/distributed.py:45: UserWarning: The dataloader, train dataloader, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 4 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  warnings.warn(*args, **kwargs)
Epoch 0:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏ | 1700/1719 [00:10<00:00, 157.72it/s, loss=0.284, v_num=16]
----------------------------
Current epoch: 0
----------------------------
Epoch 1:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏ | 1700/1719 [00:10<00:00, 164.23it/s, loss=0.200, v_num=16]
----------------------------
Current epoch: 1
----------------------------
Epoch 2:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏ | 1700/1719 [00:10<00:00, 164.02it/s, loss=0.168, v_num=16]
----------------------------
Current epoch: 2
----------------------------
Epoch 3:  10%|████████████████▊                                                                                                                                                | 180/1719 [00:01<00:09, 163.17it/s, loss=0.185, v_num=16]
^C/home/nthere/2020/pytorch-lightning/pytorch_lightning/utilities/distributed.py:45: UserWarning: Detected KeyboardInterrupt, attempting graceful shutdown...
  warnings.warn(*args, **kwargs)
Epoch 3:  10%|████████████████▊                                                                                                                                                | 180/1719 [00:01<00:10, 150.43it/s, loss=0.185, v_num=16]
```

### Expected behavior

`current_epoch` should be updated and reflect the right epoch number for the `LightningDataModule`

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
* CUDA:
        - GPU:
                - GeForce GTX 960
        - available:         True
        - version:           10.1
* Packages:
        - numpy:             1.18.5
        - pyTorch_debug:     False
        - pyTorch_version:   1.5.0+cu101
        - pytorch-lightning: 0.10.0
        - tqdm:              4.47.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.3
        - version:           #113~16.04.1-Ubuntu SMP Fri Jul 10 04:37:08 UTC 2020

```

### PR DESCRIPTION
**Pull request by nrupatunga**
## What does this PR do?

Fixes #3974

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
@ydcjeff @ananthsub @awaelchli 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -458,8 +458,14 @@ def train(self):
...
+            end_epoch = self.max_epochs
+
             # run all epochs
-            for epoch in range(self.current_epoch, self.max_epochs):
+            for epoch in range(start_epoch, end_epoch):
+
+                # Update trainer current epoch
+                self.current_epoch = epoch
 
                 # reset train dataloader
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:470]: i'd prefer to push this logic to be in `on_train_epoch_start`. then we can leave the current epoch initialization as-is.

**Review by ananthsub — CHANGES_REQUESTED** [review]
Please also add a unit test for this case. You could make sure train_dataloader is only called once with epoch=0

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (nrupatunga) [reply pytorch_lightning/trainer/trainer.py:470]: Makes sense. I will makes that change

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -459,11 +459,7 @@ def train(self):
 
         try:
             # run all epochs
-            for epoch in range(self.current_epoch, self.max_epochs):
-
-                # reset train dataloader
-                if self.reload_dataloaders_every_epoch:
-                    self.reset_train_dataloader(model)
+            for epoch in range(start_epoch, end_epoch):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:462]: this should be unchanged now
```suggestion
            for epoch in range(self.current_epoch, end_epoch):
```

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -206,6 +206,11 @@ def check_checkpoint_callback(self, should_save, is_last=False):
             [c.on_validation_end(self.trainer, model) for c in checkpoint_callbacks]
 
     def on_train_epoch_start(self, epoch):
+
+        # reset train dataloader
+        if self.trainer.reload_dataloaders_every_epoch:
+            self.trainer.reset_train_dataloader(model)
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:213]: need to swap the order to after model / current_epoch are properly initialized :)

**Review by ananthsub — CHANGES_REQUESTED** [review]

**Review by ananthsub — APPROVED** [review]

**Comment by nrupatunga** [comment]
@ananthsub I have made changes to confirm it fixes the issue. 

will try to add the unit test.

**Review by Borda — APPROVED** [review]
lgtm

**Comment by Borda** [comment]
@nrupatunga mind adding test to check that now it is correct?

**Comment by nrupatunga** [comment]
> @nrupatunga mind adding test to check that now it is correct?

sure @Borda will add the unit test now, will push the changes in a while.

File: CHANGELOG.md [hunk 4]
```diff
@@ -264,6 +264,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed for `load_from_checkpoint` ([#2776](https://github.com/PyTorchLightning/pytorch-lightning/pull/2776))
 - Fixes incorrect `batch_sizes` when Dataloader returns a dict with multiple tensors ([#3668](https://github.com/PyTorchLightning/pytorch-lightning/pull/3668))
 - Fixed unexpected signature for `validation_step` ([#3947](https://github.com/PyTorchLightning/pytorch-lightning/pull/3947))
+- Fixed `current_epoch` property update to reflect true epoch number inside `LightningDataModule`, when `reload_dataloaders_every_epoch=True`
```
Comment (ydcjeff) [inline CHANGELOG.md:267]: This changelog needs to be under [unreleased] and needs to add PR number too.

**Review by ydcjeff — APPROVED** [review]
LGTM. just changelog

File: CHANGELOG.md (hunk 4 above)
Comment (nrupatunga) [reply CHANGELOG.md:267]: My bad, somehow missed PR number while committing. Adding the same thank you.

**Comment by nrupatunga** [comment]
> @nrupatunga mind adding test to check that now it is correct?

@ydcjeff @Borda @ananthsub @awaelchli 
Can you help with the unit test? I have added a unit test, which shows what I wanted to add, But I'm not sure if that's the right way to.

**Comment by nrupatunga** [comment]
> Please also add a unit test for this case. You could make sure train_dataloader is only called once with epoch=0 with the reload every epoch flag enabled, for instance.

added unit test for the above, please review and suggest the changes.

**Pull request merged**